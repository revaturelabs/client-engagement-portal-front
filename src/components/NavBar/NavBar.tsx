import React, { useState } from "react";
import {
  Col,
  DropdownItem,
  DropdownMenu,
  DropdownToggle,
  Row,
  ButtonDropdown,
} from "reactstrap";
import logo from "../../assets/logo.png";
import "../../scss/navStyles.scss";
import { Turn as Hamburger } from "hamburger-react";
import { RootStateOrAny, useDispatch, useSelector } from "react-redux";
import { logout } from "../../actions/UserActions";
import firebase from "firebase/app";
import "firebase/auth";
import { Link, useHistory } from "react-router-dom";

/**
 * @function NavBar
 * Displays the header on any page where a user is logged in.
 *
 * @param props contains the child components that are between the opening and closing NavBar tags.
 * Should consist only of drop-down options.
 */
export const NavBar: React.FC<any> = (props: any) => {
  const [navMenuOpen, setNavMenuOpen] = useState(false);
  const [hamOpen, setHamOpen] = useState(false);
  const history = useHistory();
  /**
   * @function toggle
   * toggles the dropdown menu and the orientation of the hamburger menu
   */
  const toggle = () => {
    setNavMenuOpen(!navMenuOpen);
    setHamOpen(!navMenuOpen);
  };

  const dispatch = useDispatch();

  const user = useSelector((state: RootStateOrAny) => state.userState.user);
  const name = user
    ? user.role === "admin"
      ? `${user.firstName} ${user.lastName}`
      : user.companyName
    : "";

  /**
   * @function LogOut
   * De-authenticates the user session upon clicking the logout dropdown option.
   */
  const LogOut = () => {
    firebase.auth().signOut();
    dispatch(logout()); // clears the user data from the local state
  };

  // let logoLink = "#";
  // if (props.route) {
  //     logoLink = props.route;
  // }
  const navigateToMessages = () => {
    history.push("/messages");
  };

  return (
    <Row className="justify-content-around myNav">
      <Col xs="auto" className="justify-content-start logoContainer">
        <img id="revLogo" src={logo} className="myLogo" alt="revature logo" />
      </Col>

      <Col className="d-flex align-items-center justify-content-end auto test1">
        {/* <Link to={"/messages"}> */}
        {/* <a href="/messages" className="btn btn-primary mr-2">
          Messages
        </a> */}
        {/* </Link> */}
        <button className="btn btn-primary mr-2" onClick={navigateToMessages}>
          Messages
        </button>
        <ButtonDropdown isOpen={navMenuOpen} toggle={toggle}>
          {/* Mobile Hamburger Menu */}
          <DropdownToggle
            id="navDropButton"
            style={{ margin: "10px", backgroundColor: "white", border: "none" }}
          >
            <span
              className="myDropdown"
              id="usersName"
              style={{
                margin: "5px",
                color: "#474C55",
                backgroundColor: "white",
                border: "none",
              }}
            >
              Welcome {name} &#9660;
            </span>
            <span className="myMobileDropdown">
              <Hamburger
                hideOutline={true}
                toggled={hamOpen}
                toggle={setHamOpen}
                color="#474C55"
              ></Hamburger>
            </span>
          </DropdownToggle>
          {/* Desktop Menu */}
          <DropdownMenu right>
            {props.children}
            <DropdownItem id="test-head" header>
              Account Options
            </DropdownItem>
            <a href="/login">
              <DropdownItem onClick={LogOut}>Logout</DropdownItem>
            </a>
          </DropdownMenu>
        </ButtonDropdown>
      </Col>
    </Row>
  );
};
