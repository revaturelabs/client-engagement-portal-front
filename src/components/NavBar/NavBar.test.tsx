import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { NavBar } from './NavBar';
import { mount } from "enzyme";
import { ButtonDropdown } from 'reactstrap';
import thunk from 'redux-thunk';
import logo from '../../assets/logo.png';

let store:any;
let component:any;
let wrapper:any;

const mockStore = configureStore([ thunk ]);

describe('NavBar component', () => {

    test('should render with given state from Redux store', ()=>{
        store = mockStore({
            userState:{
                user:{
                    firstName: 'wack',
                    lastName: 'jobs',
                    role: 'admin'
                }     
            }   
        });
        component = renderer.create(
            <Provider store={store}>
               <NavBar route="test" logoSrc={ logo }/>
           </Provider>
       );

        expect(component.toJSON()).toMatchSnapshot();
    });

    test("dropdown menu can toggle on and off", () => {

        store = mockStore({
            userState:{
                user:{
                    firstName: 'wack',
                    lastName: 'jobs',
                    companyName: 'wally world',
                    role:'client'
                }     
            }   
        }); 
    
        wrapper = mount(
            <Provider store={store}>
               <NavBar  route="test" logoSrc={ logo }/>
           </Provider> 
       );

        expect(wrapper.find(ButtonDropdown).prop("isOpen")).toBe(false);
        wrapper.find("button#navDropButton").simulate("click");
        expect(wrapper.find(ButtonDropdown).prop("isOpen")).toBe(true);
    });
});
