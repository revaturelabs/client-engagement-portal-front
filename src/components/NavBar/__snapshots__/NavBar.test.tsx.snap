// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar component should render with given state from Redux store 1`] = `
<div
  className="justify-content-around myNav row"
>
  <div
    className="justify-content-start logoContainer col-auto"
  >
    <img
      alt="revature logo"
      className="myLogo"
      id="revLogo"
      src=""
    />
  </div>
  <div
    className="d-flex align-items-center justify-content-end auto test1 col"
  >
    <div
      className="btn-group"
      onKeyDown={[Function]}
    >
      <button
        aria-expanded={false}
        aria-haspopup={true}
        aria-label={null}
        className="btn btn-secondary"
        id="navDropButton"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "border": "none",
            "margin": "10px",
          }
        }
        type="button"
      >
        <span
          className="myDropdown"
          id="usersName"
          style={
            Object {
              "backgroundColor": "white",
              "border": "none",
              "color": "#474C55",
              "margin": "5px",
            }
          }
        >
          Welcome 
          wack jobs
           â–¼
        </span>
        <span
          className="myMobileDropdown"
        >
          <div
            className="hamburger-react"
            onClick={[Function]}
            onKeyUp={[Function]}
            role="button"
            style={
              Object {
                "cursor": "pointer",
                "height": "48px",
                "outline": "none",
                "position": "relative",
                "transition": "0.4s cubic-bezier(0, 0, 0, 1)",
                "userSelect": "none",
                "width": "48px",
              }
            }
            tabIndex={0}
          >
            <div
              style={
                Object {
                  "background": "#474C55",
                  "height": "3px",
                  "left": "8px",
                  "position": "absolute",
                  "top": "13px",
                  "transform": "none",
                  "transition": "0.4s cubic-bezier(0, 0, 0, 1)",
                  "width": "32px",
                }
              }
            />
            <div
              style={
                Object {
                  "background": "#474C55",
                  "height": "3px",
                  "left": "8px",
                  "position": "absolute",
                  "top": "23px",
                  "transform": "none",
                  "transition": "0.2s cubic-bezier(0, 0, 0, 1)",
                  "width": "32px",
                }
              }
            />
            <div
              style={
                Object {
                  "background": "#474C55",
                  "height": "3px",
                  "left": "8px",
                  "position": "absolute",
                  "top": "33px",
                  "transform": "none",
                  "transition": "0.4s cubic-bezier(0, 0, 0, 1)",
                  "width": "32px",
                }
              }
            />
          </div>
        </span>
      </button>
      <div
        aria-hidden={true}
        className="dropdown-menu dropdown-menu-right"
        role="menu"
        tabIndex="-1"
      >
        <h6
          className="dropdown-header"
          id="test-head"
          onClick={[Function]}
          tabIndex="-1"
        >
          Account Options
        </h6>
        <a
          href="/login"
        >
          <button
            className="dropdown-item"
            onClick={[Function]}
            role="menuitem"
            tabIndex="0"
            type="button"
          >
            Logout
          </button>
        </a>
      </div>
    </div>
  </div>
</div>
`;
